airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/Destination.java

```diff
@@ -29,8 +35,129 @@ AirbyteMessageConsumer getConsumer(JsonNode config,
+   */
+  @Slf4j
+  class ShimToSerializedAirbyteMessageConsumer implements SerializedAirbyteMessageConsumer {
+
+    private final AirbyteMessageConsumer consumer;
+
+    public ShimToSerializedAirbyteMessageConsumer(final AirbyteMessageConsumer consumer) {
+      this.consumer = consumer;
+    }
+
+    @Override
+    public void start() throws Exception {
+      consumer.start();
+    }
+
+    /**
+     * Consumes an {@link AirbyteMessage} for processing.
+     * <p>
+     * If the provided JSON string is invalid AND represents a {@link AirbyteMessage.Type#STATE}
+     * message, processing is halted. Otherwise, the invalid message is logged and execution continues.
+     *
+     * @param inputString JSON representation of an {@link AirbyteMessage}.
+     * @throws Exception if an invalid state message is provided or the consumer is unable to accept the
+     *         provided message.
+     */
+    @Override
+    public void accept(final String inputString, final Integer sizeInBytes) throws Exception {
+      consumeMessage(consumer, inputString);
+    }
+
+    @Override
+    public void close() throws Exception {
+      consumer.close();
+    }
+
+    /**
+     * Consumes an {@link AirbyteMessage} for processing.
+     * <p>
+     * If the provided JSON string is invalid AND represents a {@link AirbyteMessage.Type#STATE}
+     * message, processing is halted. Otherwise, the invalid message is logged and execution continues.
+     *
+     * @param consumer An {@link AirbyteMessageConsumer} that can handle the provided message.
+     * @param inputString JSON representation of an {@link AirbyteMessage}.
+     * @throws Exception if an invalid state message is provided or the consumer is unable to accept the
+     *         provided message.
+     */
+    @VisibleForTesting
+    static void consumeMessage(final AirbyteMessageConsumer consumer, final String inputString) throws Exception {
+
+      final Optional<AirbyteMessage> messageOptional = Jsons.tryDeserialize(inputString, AirbyteMessage.class);
+      if (messageOptional.isPresent()) {
+        consumer.accept(messageOptional.get());
+      } else {
+        if (isStateMessage(inputString)) {
+          throw new IllegalStateException("Invalid state message: " + inputString);
+        } else {
+          log.error("Received invalid message: " + inputString);
+        }
+      }
+    }
+
+    /**
+     * Tests whether the provided JSON string represents a state message.
+     *
+     * @param input a JSON string that represents an {@link AirbyteMessage}.
+     * @return {@code true} if the message is a state message, {@code false} otherwise.
+     */
+    @SuppressWarnings("OptionalIsPresent")
+    private static boolean isStateMessage(final String input) {
+      final Optional<AirbyteTypeMessage> deserialized = Jsons.tryDeserialize(input, AirbyteTypeMessage.class);
```

Comment:
I'm sure you tested this, but just for my own understanding, how does this work?  Since there's only one top-level property, with the JSON parser only deser the top level, leaving `RECORD: {}` as a string?

Reply:
cc @davinchia remember when we tried this in mexico with just `string.contains()`? this is much better

Reply:
fwiw, jonathan pearlin is the one who set this up. this has been in prod for a while.